mobile/src/main/java/ch/epfl/sdp/mobile/application/chess/engine/rules/Moves.kt

```diff
@@ -36,19 +39,51 @@ typealias Moves<P> = Sequence<Pair<Action, Effect<P>>>
  * @param x the delta on the x axis.
  * @param y the delta on the y axis.
  * @param includeAdversary true if the [Moves] should include stepping on and eating adversaries.
+ * @param promotionAllowed true if the [Moves] should include promotion of the piece.
  */
 fun BoardSequence<Piece<Role>>.delta(
     from: Position,
     x: Int,
     y: Int,
     includeAdversary: Boolean = true,
+    promotionAllowed: Boolean = false,
 ): Moves<Piece<Role>> = sequence {
   val board = first()
   val delta = Delta(x, y)
   val target = from + delta ?: return@sequence // Stop if out of bounds.
   val piece = board[target]
   if (piece == null || (includeAdversary && piece.color == Adversary)) {
-    yield(Action(from, delta) to move(from, delta))
+    yieldMoveOrPromote(board, from, target, promotionAllowed = promotionAllowed)
+  }
+}
+
+/**
+ * Yields a [Move] or a [Promote] action, depending on the value of the [promote] argument.
+ *
+ * @param Color the color of the pieces.
+ * @param board the current [Board] state.
+ * @param from the start [Position].
+ * @param to the target [Position].
+ * @param promotionAllowed true iff the promotion action is allowed.
+ */
+private suspend fun <Color> SequenceScope<Pair<Action, Effect<Piece<Color>>>>.yieldMoveOrPromote(
+    board: Board<Piece<Color>>,
+    from: Position,
+    to: Position,
+    promotionAllowed: Boolean,
+) {
+  val isOnLastRow = to.y == 0
+  if (!isOnLastRow || !promotionAllowed) {
+    yield(Move(from, to) to move(from, to))
+  } else {
+    // Retrieve the origin piece.
+    val piece = board[from] ?: return
+    // Compute the identifier of the promoted piece.
+    val id = board.maxOf { (_, piece) -> piece.id }.inc()
```

Comment:
Is there a reason you need to compute a new identifier for the promoted piece? Could you not just reuse the pawn's identifier, since it will be completely replaced?

Reply:
Fixed in 0f09f58d9ca27f5e48bbe259823344e119396855.